# Regressão 01 - tarefa 02 - transformações em X e Y

### 1. Melhorando o ajuste no percentual de gorjetas

a. Vamos partir do modelo feito na última tarefa, o que relaciona ```tip_pct``` e ```net_bill```. Carregue a base, os pacotes e reproduza este modelo aqui.

In [23]:
import pandas as pd
import seaborn as sns
from seaborn import load_dataset

import matplotlib.pyplot as plt

import numpy as np

import statsmodels.formula.api as smf

In [24]:
tips = sns.load_dataset("tips")
tips['tip_pct'] = tips['tip'] / (tips['total_bill'] - tips['tip'])
tips['net_bill'] = tips['total_bill'] - tips['tip']
tips.head()

total_bill   tip     sex smoker  day    time  size   tip_pct  net_bill
0       16.99  1.01  Female     No  Sun  Dinner     2  0.063204     15.98
1       10.34  1.66    Male     No  Sun  Dinner     3  0.191244      8.68
2       21.01  3.50    Male     No  Sun  Dinner     3  0.199886     17.51
3       23.68  3.31    Male     No  Sun  Dinner     2  0.162494     20.37
4       24.59  3.61  Female     No  Sun  Dinner     4  0.172069     20.98

### 2. ajuste o modelo de ```tip_pct``` em função do logaritmo de ```net_bill```.

In [25]:
# Calculando o logaritmo do valor conta liquido
tips['log_net_bill'] = np.log(tips['net_bill'])

# Ajustando a regressão de tip_pct por log_net_bill
reg_tip_pct = smf.ols(formula='tip_pct ~ log_net_bill', data=tips).fit()

reg_tip_pct.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.198
Method:                 Least Squares   F-statistic:                     61.16
Date:                Sun, 09 Jun 2024   Prob (F-statistic):           1.62e-13
Time:                        15:47:08   Log-Likelihood:                 123.81
No. Observations:                 244   AIC:                            -243.6
Df Residuals:                     242   BIC:                            -236.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept        0.6181      0.054     11.444      0.000       0.512       0.724
log_net_bill    -0.1534      0.020     -7.820      0.000      -0.192      -0.115
==============================================================================
Omnibus:                      440.899   Durbin-Watson:                   1.965
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           171509.052
Skew:                           9.850   Prob(JB):                         0.00
Kurtosis:                     131.381   Cond. No.                         17.9
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

###  3. ajuste o modelo de ```tip_pct``` em função de um polinômio de segundo grau em ```net_bill```.

In [26]:
# Criando colunas de polinomio de 2º em net_bill
tips['net_bill_squared'] = np.power(tips['net_bill'], 2)

reg_tip_pct_poly = smf.ols(formula='tip_pct ~net_bill +  net_bill_squared', data=tips).fit()

reg_tip_pct_poly.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                tip_pct   R-squared:                       0.158
Model:                            OLS   Adj. R-squared:                  0.151
Method:                 Least Squares   F-statistic:                     22.55
Date:                Sun, 09 Jun 2024   Prob (F-statistic):           1.06e-09
Time:                        15:47:39   Log-Likelihood:                 117.25
No. Observations:                 244   AIC:                            -228.5
Df Residuals:                     241   BIC:                            -218.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept            0.4851      0.047     10.332      0.000       0.393       0.578
net_bill            -0.0268      0.005     -5.477      0.000      -0.036      -0.017
net_bill_squared     0.0005      0.000      4.356      0.000       0.000       0.001
==============================================================================
Omnibus:                      447.620   Durbin-Watson:                   2.026
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           186035.672
Skew:                          10.140   Prob(JB):                         0.00
Kurtosis:                     136.744   Cond. No.                     2.42e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.42e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

###  4. ajuste um modelo no log do ```tip_pct``` em função de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```

In [28]:
# Criando coluna com logaritmo do tip_pct
tips['log_tip_pct'] = np.log(tips['tip_pct'])

# Ajustando a regressão do log_tip_pct por net_bill
reg_log_tip_pct = smf.ols(formula='log_tip_pct ~net_bill', data=tips).fit()

# Calculando o R² do modelo em tip_pct
r_tip_pct = reg_log_tip_pct.rsquared

print(f"R² do modelo em tip_pct: {r_tip_pct:.4f}")

R² do modelo em tip_pct: 0.2353


###  5. ajuste um modelo no log do ```tip_pct``` em função do log de ```net_bill```. Calcule o $R^2$ também em ```tip_pct```.

In [29]:
# Ajustando a regressão do log_tip_pct por log_net_bill
reg_log_tip_pct_log_net_bill = smf.ols(formula='np.log(tip_pct) ~np.log(net_bill)', data=tips).fit()

# Calculando R² do modelo em tip_pct
r_tip_pct_log_net_bill = reg_log_tip_pct_log_net_bill.rsquared

print(f"R² do MOdelo em tip_pct: {r_tip_pct_log_net_bill:.4f}")

R² do MOdelo em tip_pct: 0.2935


###  6. qual modelo resultou no melhor ajuste?

O último modelo no log do tip_pct em função do log de net_bill com 0.2935